tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,506 @@
+/*
+    COPYRIGHT
+
+    All contributions by Taiga Nomi
+    Copyright (c) 2013, Taiga Nomi
+    All rights reserved.
+
+    All other contributions:
+    Copyright (c) 2013-2016, the respective contributors.
+    All rights reserved.
+
+    Each contributor holds copyright over their respective contributions.
+    The project versioning (Git) records all such contribution source information.
+
+    LICENSE
+
+    The BSD 3-Clause License
+
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright notice, this
+      list of conditions and the following disclaimer.
+
+    * Redistributions in binary form must reproduce the above copyright notice,
+      this list of conditions and the following disclaimer in the documentation
+      and/or other materials provided with the distribution.
+
+    * Neither the name of tiny-dnn nor the names of its
+      contributors may be used to endorse or promote products derived from
+      this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+    SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#pragma once
+
+#include <cmath> // sqrt
+#include <algorithm> // std::fill, std::generate
+
+#include "tiny_dnn/core/framework/device.fwd.h"
+
+namespace tiny_dnn {
+
+template<typename U = float_t>
+class Tensor {
+ public:
+    /*
+     * Initializes an empty tensor.
+     */
+    Tensor() {}
+
+    /*
+     * Create a tensor of the given dimension.
+     * It is assumed that a tensor will hold data in NxWxHxD order,
+     * where:
+     *  N the batch axis
+     *  W the width axis
+     *  H the heigth axis
+     *  D the depth axis
+     *
+     *  Data will be hold by a std::vector with 64bytes alignment.
+     */
+    explicit Tensor(const cnn_size_t d0,
```

Comment:
ok, but before that: are we going to support tensors serialization? (could be cool this feature :D)  If so, I think that `cnn_size_t` a.k.a `std::uint32_t` should be okay. Otherwise, I'll move to standard `size_t`.